File: android/src/main/kotlin/com/app/base_plugin/BasePlugin.kt
package com.app.base_plugin

import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import greeting.Greeting
import greeting.WalletResponse
import org.json.JSONObject

fun WalletResponse.toJSONObject(): JSONObject {
    val jsonObject = JSONObject()
    jsonObject.put("address", address)
    jsonObject.put("mnemonic", mnemonic)
    jsonObject.put("privateKey", privateKey)
    jsonObject.put("publicKey", publicKey)
    jsonObject.put("keystore", keystore)
    return jsonObject
}

/** BasePlugin */
class BasePlugin : FlutterPlugin, HostMessageApi {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var flutterMessageApi: FlutterMessageApi
    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        print("onAttachedToEngine")
        HostMessageApi.setUp(flutterPluginBinding.binaryMessenger, this)
        flutterMessageApi = FlutterMessageApi(flutterPluginBinding.binaryMessenger)
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        HostMessageApi.setUp(binding.binaryMessenger, null)
    }

    override fun flutter2Native(message: String, type: Long): String {
        print("flutter2Native=message=start=>$message=type=$type")
        flutterMessageApi.native2Flutter("input params=>2222", callback = {
            print("response=>${it.getOrNull()}")
        }, typeArg = 1)
        when (type) {
            1L -> {
                val wallet = Greeting.createWalletF1("111111")
                return wallet.toJSONObject().toString()
            }
            2L -> {
                return Greeting.sayHi("flutter2Native=message=$message=type=$type")
            }

            else -> {
                return "flutter2Native=message=$message=type=$type"
            }
        }
    }

    override fun flutter2NativeAsync(
        message: String,
        type: Long,
        callback: (Result<String>) -> Unit
    ) {
        fun d(e: Result<String>) {
            print("d")
        }
        print("flutter2NativeAsync=message=$message=type=$type")
        flutterMessageApi.native2FlutterAsync("2222", callback = {
            print("222")
        }, typeArg = 2)
        callback(Result.success("flutter2NativeAsync=message=$message=type=$type"));
    }
}

File: ios/Classes/BasePlugin.swift
import Flutter
import UIKit
import Greeting
public class BasePlugin: NSObject, FlutterPlugin,HostMessageApi {
    
    private static var flutterAPI: FlutterMessageApi? = nil
    public static func register(with registrar: any FlutterPluginRegistrar) {
        let api = BasePlugin()
        HostMessageApiSetup.setUp(binaryMessenger: registrar.messenger(), api: api)
        flutterAPI = FlutterMessageApi(binaryMessenger: registrar.messenger())
    }
    
    func flutter2Native(message: String, type: Int64) throws -> String {
        print("ios->flutter2Native=>start=>message=\(message)=type=\(type)");
        if type==1 {
            var error: NSError?
            if let walletResponse = GreetingCreateWalletF1( "111111", &error) {
               let dict: [String: Any] = [
                   "address": walletResponse.address,
                   "privateKey": walletResponse.privateKey,
                   "publicKey": walletResponse.publicKey,
                   "mnemonic": walletResponse.mnemonic,
                   "keystore": walletResponse.keystore
               ]
               do {
                   let jsonData = try JSONSerialization.data(withJSONObject: dict, options: .prettyPrinted)
                   let jsonString = String(data: jsonData, encoding: .utf8) ?? ""
                    // 在这里处理成功创建钱包后的walletResponse，比如获取钱包相关信息等
                   return jsonString
               } catch {
                   print("Error converting to JSON: \(error)")
               }
            return ""
            } else {
                // 在这里处理创建钱包失败的情况，通过error获取错误信息
                if let unwrappedError = error {
                    print("创建钱包失败，错误信息：\(unwrappedError.localizedDescription)")
                } else {
                    print("创建钱包失败，但未获取到具体错误信息")
                }
                return ""
            }
        } else if type==2{
            return "这是日志信息"+GreetingSayHi("ios->flutter2Native=>start=>message=\(message)=type=\(type)")
        }else{
            return "ios->flutter2Native=>start=>message=\(message)=type=\(type)"
        }
    }


    func flutter2NativeAsync(message: String, type: Int64, completion: @escaping (Result<String, any Error>) -> Void) {
        print("ios->flutter2NativeAsyncMessage===2")
        completion(.success(message))
    }

    

}

